### DISCUSSION TITLE
CRD to manage argocd resources

### DISCUSSION THREAD

**Discussion by nissessenap** [argoproj/argo-cd#13142] [Ideas]
## Summary

I would like to have repository CRD to enable developers to add repositories in a more Kubernetes native way, then using secrets that you annotate.

## Background

Today ArgoCD has 3 CRDs

- project
- application
- applicationsets

In general, you can say that they are used to managing applications.

In general, a ArgoCD admin don't create multiple argocd instances so integrating the ArgoCD settings isn't super prioritized.

But add a new set of CRDs to enable normal users (if you want) to add new **repositories** feels like a natural part of Argo.
Same thing could also be useful for **GnuPG keys** and **Certificates**, but they are not as important compared to repositories since how often it's done.

## Why?

Adding one or multiple CRDs for repositories would enable us to verify the input that the users do in a whole other way.

It could also enable us to manage things like OCI resources through another CRD if that is something that we think is a good idea.

It's just easier for people that are used to controllers.

## Backwards compatibility

Since we don't want to make breaking changes, we can start with implementing the controllers and just use them to generate secrets according to the current standard.
If/when ArgoCD community it's time to do breaking changes and release a 3.0 we will be able to remove the old secret solution.
This will enable us to remove lots of logic that the CRD will have built in vs the parsing of the secrets.

## Forward

I'm new to the ArgoCD community, but I see that there is a proposals system in place.
If the community thinks this is a good idea, I can try to write a basic proposal on how the CRDs could look like and take it from there.

**Comment by blakepettersson** [comment]
For some historical context: #5436 and #6103

**Comment by jgwest** [comment]
This would be the relevant issue, where this was proposed: https://github.com/argoproj/argo-cd/issues/5720

**Comment by nissessenap** [comment]
We had a discussion around the during today's maintainers meeting, and I got positive feedback around this.
So I will start working on the API how it could look like in a proposal and continue the discussion there.

Then we will most likely create a separate controller and store it in `argoproj-labs` git org for starters and continue from there.

If anyone has thoughts on this already, feel free to reach out, and I'm happy to work on this together.

**Reply by jannfis** [reply]
I unfortunately couldn't make it to the contributor's meeting this week, but I was wondering why a separate controller would be necessary for this or what has been the discussions around this?

Personally, I'm all for having clusters and repositories as CRDs being a first-class feature, directly integrated into `argocd-server` and `argocd-application-controller` components.
